### PR TITLE
Add version management code

### PR DESCRIPTION
**Pull request by kyleschlosser**
This pull request adds reconciliation logic to the Kabanero instance `version` attribute: 

```
apiVersion: kabanero.io/v1alpha1
kind: Kabanero
metadata:
  name: kabanero
spec:
  version: 0.1.2       <-- here
```

Prior to this pull request, regardless of the value found in this attribute, the operator would always apply the same orchestration logic. 

In the process of implementing this pull request several changes where made: 
- A mapping document between Kabanero versions and managed software versions was created
- The model was enhanced to include appsody, landing page, and cli services elements
- When possible, version and image overrides where added to the model as well as to the reconciliation logic

### CODE REVIEW

File: config/versions.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+# Note the following sequence: the Kabanero operator becomes aware
+# of a new version of the Kabanero platform, in part by editing this file
+# Once the operators are rolled out, the administrator can choose to use
+# the new Kabanero version by editing the Kabanero resource. This causes
+# other resources to be updated using this mapping
+
+# When the Kabanero instance does not specify which version of Kabanero to use,
+# this is the value
+default: "0.1.2"
```
[the remark this answers is not in the document]
Comment (kyleschlosser) [reply config/versions.yaml]: The default should not be treated as the indicator of what the current release is, but rather what release you want the majority of customers to run (those which have not pinned their own versions). If you update the default to 0.1.3 now, but the dependencies are not yet available, running a build on master without a `version: 0.1.2` attribute set would mean that the default configuration is broken (references non-existing image tags). 

Updating the default in lock step with the availability of a new release gives a behavior similar to what we had before - installing an operator update will automatically update the Kabanero instances the operator is controlling. Ideally, updates to the default should be judicious, not based upon what's new but rather what you want most customers running. 

Let's say you had a 0.1.2, 0.1.3, 0.1.4 in the field, and you just release 1.0.0 which you are concerned may have breaking changes. Maybe you leave the default at 0.1.4, and wait until 1.0.1 is available before changing the default to 1.0.1, this way some of the kinks in the 1.0.0 release can get worked out. The default is never set to 1.0.0 in this case, it goes straight from 0.1.4 to 1.0.1.

File: pkg/assets/config/config_test.go [hunk 2]
```diff
@@ -5,12 +5,14 @@ import (
 )
 
 func TestAssets(t *testing.T) {
+	path := "orchestrations/appsody-operator/0.1/appsody.yaml"
```
[the remark this answers is not in the document]
Comment (kyleschlosser) [reply pkg/assets/config/config_test.go]: Yes, it is important to understand the orchestrations (configuration yaml) and images can change at a different rate. So until the Appsody team says they need some other yaml, we keep that in place and continue changing the image on each release. Although its easier to comprehend when we keep these concepts 1-1 (yaml and images are provided together as a bundle for every release), this means that for N releases of a dependency you also have N copies of the yaml to manage. 

One implication to what we have here is that Appsody cannot change the yaml in their own patch release, they have to wait for the next operator patch release.

File: pkg/apis/kabanero/v1alpha1/kabanero_types.go [hunk 3]
```diff
@@ -47,6 +53,26 @@ type KappnavCustomizationSpec struct {
...
+type AppsodyCustomizationSpec struct {
+	Enable     bool   `json:"enable,omitempty"`
+	Version    string `json:"version,omitempty"`
+	Image      string `json:"image,omitempty"`
+	Repository string `json:"repository,omitempty"`
+	Tag        string `json:"tag,omitempty"`
+}
+
+type KabaneroCliServicesCustomizationSpec struct {
+	Enable     bool   `json:"enable,omitempty"`
```
[the remark this answers is not in the document]
Comment (kyleschlosser) [reply pkg/apis/kabanero/v1alpha1/kabanero_types.go]: You are correct, we need to get a defaulter in place to correct this, or we have to rewrite all of our configuration attributes based upon what we want the default to be.

File: config/versions.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaczyns) [reply config/versions.yaml]: OK, I understand.  So I think then that we need to make it clearer to a PR creator, which version yaml they should be updating to be included in the next Kabanero operator version.

File: pkg/assets/config/config_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kaczyns) [reply pkg/assets/config/config_test.go]: Thanks for explaining that.

File: pkg/apis/kabanero/v1alpha1/kabanero_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaczyns) [reply pkg/apis/kabanero/v1alpha1/kabanero_types.go]: I'm not sure what a defaulter is, but I was hoping that there was a way in the json annotation to say what we want the default to be?  Is that what a defaulter is?

[the remark this answers is not in the document]
Comment (kaczyns) [reply pkg/apis/kabanero/v1alpha1/kabanero_types.go]: Thank you for removing enabled on the landing page.  We can come back and fix the defaulting later.

**Pull request merged**